### ISSUE TITLE
Git Town doesn't delete merged branches when running `git town sync`

### ISSUE THREAD

**Issue by stephenwade** [git-town/git-town#4222]
### Steps to reproduce

1. Create a branch and `propose` it
2. Merge the proposal in GitHub
3. Run `git town sync -a`

I expected Git Town to delete the branch, but it didn't.

### Command output

I didn't use `--verbose` because I'm filing the report after running into the issue. If necessary, I can use `--verbose` next time I'm in an equivalent situation and report back.

```shell
❯ git hack hooks

[main] git fetch --prune --tags

[main] git rebase origin/main --no-update-refs
Current branch main is up to date.

[main] git checkout -b hooks
Switched to a new branch 'hooks'

❯ # create files in VS Code
zsh: command not found: #

❯ git add .

❯ git commit -m 'Add hooks'
[hooks 4b42a38] Add hooks
 3 files changed, 19 insertions(+)
 create mode 100644 .husky/pre-merge-commit
 create mode 100644 .husky/pre-push

❯ git town propose

[hooks] git fetch --prune --tags

Looking for proposal online ... none

[hooks] git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

[main] git rebase origin/main --no-update-refs
Current branch main is up to date.

[main] git checkout hooks
Switched to branch 'hooks'

[hooks] git rebase main --no-update-refs
Current branch hooks is up to date.

[hooks] git push -u origin hooks
remote: 
remote: Create a pull request for 'hooks' on GitHub by visiting:        
remote:      https://github.com/MYORG/MYREPO/pull/new/hooks        
remote: 
To https://github.com/MYORG/MYREPO.git
 * [new branch]      hooks -> hooks
branch 'hooks' set up to track 'origin/hooks'.

open https://github.com/MYORG/MYREPO/compare/hooks?expand=1


branch "hooks" is no longer a prototype branch

❯ # merge in GitHub
zsh: command not found: #

❯ git town sync -a

[hooks] git fetch --prune --tags

[hooks] git checkout main
Switched to branch 'main'
Your branch is behind 'origin/main' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)

[main] git rebase origin/main --no-update-refs
Successfully rebased and updated refs/heads/main.

[main] git checkout hooks
Switched to branch 'hooks'
Your branch is based on 'origin/hooks', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)

[hooks] git push --tags
Everything up-to-date

❯ git branch --show-current
hooks
```

### Additional information

I think this used to work as expected, but I'm not sure what version that was. Let me know if you need me to bisect.

**Comment by kevgo** [comment]
Thanks for reporting this!

It looks like we already have an [end-to-end test](https://github.com/git-town/git-town/blob/main/features/sync/current_branch/feature_branch/merge_sync_strategy/tracking_branch/deleted/happy_path.feature) that covers this scenario, albeit without the `--all` flag. Since the flag shouldn’t affect the outcome, this test indirectly validates the behavior.

To double-check, I created a more targeted [end-to-end test](https://github.com/git-town/git-town/blob/main/features/sync/all_branches/merge_sync_strategy/ship_prototype_branch_and_sync.feature) that replicates the exact scenario, including the --all flag. Both tests pass on my end.

Given that, I’m having trouble reproducing the issue. Could you share more details about your setup or any additional steps that might help pinpoint the problem?

**Comment by stephenwade** [comment]
Hmm, I don't think I'm doing anything unusual. Here's my `.git-branches.toml` config file:

```toml
create-prototype-branches = true
ship-delete-tracking-branch = false
sync-tags = false

[branches]
main = "main"
perennials = ["production"]

[sync-strategy]
prototype-branches = "rebase"
```

**Comment by stephenwade** [comment]
I have an idea of what might be different. My repo uses squash merging, so the commit on `main` doesn't have the same hash as the commit on the branch `hooks`.

**Comment by kevgo** [comment]
> It's not clear to me what exactly the E2E test does when you write
> ```
> And origin ships the "hooks" branch
> ```

Great feedback! I have made this step more descriptive. It now looks like this:

```
And origin ships the "feature" branch using the "squash-merge" ship-strategy
```

**Comment by kevgo** [comment]
You can see which code each step executes in the [step definitions file](https://github.com/git-town/git-town/blob/main/test/cucumber/steps.go#L1004-L1022). In this case, the origin remote squash-merges the branch into main. It happens on these lines:

```go
originRepo.CheckoutBranch("main")
err = originRepo.SquashMerge(originRepo.TestRunner, branchToShip)
```

**Comment by kevgo** [comment]
In your Git Town config file, you could move this line:

```toml
create-prototype-branches = true
```

Into the `[branches]` section and change it to this:

```toml
[branches]
default-type = "prototype"
```

More info at https://www.git-town.com/preferences/default-branch-type#configure-in-config-file.

**Comment by kevgo** [comment]
If you’re able to bisect the problem to identify the commit that introduced the issue, that would be a big help! You can do this automatically using the following steps:

1. Add `@this` above the first line of [the end-to-end test that reproduces your use case](https://github.com/git-town/git-town/blob/main/features/sync/all_branches/merge_sync_strategy/ship_prototype_branch_and_sync.feature). For more details, see the [documentation on end-to-end tests](https://github.com/git-town/git-town/blob/main/docs/DEVELOPMENT.md#end-to-end-tests).
2. Run `make cukethis`. This will:
   - Compile and install Git Town from source on your machine.
   - Execute the end-to-end test using the freshly compiled binary.
   - Note: Ensure that `~/go/bin` is included in your `$PATH`.
3. Use `make cukethis` as the [bisect run script](https://git-scm.com/docs/git-bisect#_bisect_run) to automate the bisect process.

Let me know if you run into any issues or have questions!

**Comment by stephenwade** [comment]
I reproduced the issue while running `git town sync -a --verbose`. The log is below.

I'll work on bisecting soon.

<details>
<summary>Command output</summary>

```
❯ git town sync -a --verbose
Git Town 16.6.0

(verbose) git version
git version 2.47.0

(verbose) git rev-parse --show-toplevel
/Users/stephen.wade/git/REDACTED

(verbose) git config -lz --includes --global
user.name
Stephen Wadeuser.email
REDACTEDpull.rebase
truepull.ff
onlypush.autosetupremote
truediff.colormoved
zebracore.pager
less -x2core.fsmonitor
trueinit.defaultbranch
mainfetch.prune
truerebase.updaterefs
truegit-town.github-token
REDACTEDalias.append
town appendalias.compress
town compressalias.contribute
town contributealias.diff-parent
town diff-parentalias.hack
town hackalias.observe
town observealias.park
town parkalias.prepend
town prependalias.propose
town proposealias.repo
town repoalias.set-parent
town set-parentalias.ship
town shipalias.sync
town syncalias.ls
town branchalias.co
town switchalias.delete
town deletealias.rename
town renamemaintenance.repo
/Users/stephen.wade/git/REDACTED
(verbose) git config -lz --includes --local
core.repositoryformatversion
0core.filemode
truecore.bare
falsecore.logallrefupdates
truecore.ignorecase
truecore.precomposeunicode
truecore.hookspath
.husky/_remote.origin.url
https://github.com/REDACTED/REDACTED.gitremote.origin.fetch
+refs/heads/*:refs/remotes/origin/*branch.main.remote
originbranch.main.merge
refs/heads/mainbranch.main.vscode-merge-base
origin/maingit-town.github-token
REDACTEDgit-town.observed-branches
mui/base-ui/9 type-check-allgit-town.prototype-branches
test iovalkey new-iconifygit-town-branch.mui/base-ui/9.parent
mainlog.excludedecoration
refs/gitbutlerbranch.production.remote
originbranch.production.merge
refs/heads/productionbranch.production.vscode-merge-base
origin/mainmaintenance.auto
falsemaintenance.strategy
incrementalbranch.mui/base-ui/9.vscode-merge-base
origin/maingit-town-branch.router.parent
mainbranch.router.vscode-merge-base
origin/mainbranch.router.remote
originbranch.router.merge
refs/heads/router
(verbose) git branch -vva --sort=refname
  main                                                               9f4728e5 [origin/main: behind 1] Set `sync-tags = false` in Git Town settings (#559)
  mui/base-ui/9                                                      b3c47700 mui/material-ui#41031
  production                                                         cc2389b9 [origin/production] Merge branch 'main' into production
* router                                                             908abe83 [origin/router: gone] Split router and root providers into separate files
  remotes/origin/HEAD                                                -> origin/main
  remotes/origin/client-frontend-temp                                ac6ee219 temp
  remotes/origin/main                                                e0383e6f Split router and root providers into separate files (#561)
  remotes/origin/nx-infer/5-webpack                                  f395d079 Use inferred tasks for Webpack
  remotes/origin/production                                          cc2389b9 Merge branch 'main' into production
  remotes/origin/renovate/non-major-dependencies                     ac497e6e chore(deps): update non-major dependencies
  remotes/origin/testing                                             a546945d Cherry-pick bug fixes from main (#204)
  remotes/origin/update-now-playing-when-channel-or-playlist-changes 67ffaaf7 wip: Trying to figure this out.
  remotes/origin/user-avatar                                         68b0e61c start upload

(verbose) git status --long --ignore-submodules
On branch router
Your branch is based on 'origin/router', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)

nothing to commit, working tree clean

(verbose) git remote
origin

[router] git fetch --prune --no-tags

(verbose) git stash list
stash@{0}: WIP on (no branch): 54289567 Add test for alwaysMask

(verbose) git branch -vva --sort=refname
  main                                                               9f4728e5 [origin/main: behind 1] Set `sync-tags = false` in Git Town settings (#559)
  mui/base-ui/9                                                      b3c47700 mui/material-ui#41031
  production                                                         cc2389b9 [origin/production] Merge branch 'main' into production
* router                                                             908abe83 [origin/router: gone] Split router and root providers into separate files
  remotes/origin/HEAD                                                -> origin/main
  remotes/origin/client-frontend-temp                                ac6ee219 temp
  remotes/origin/main                                                e0383e6f Split router and root providers into separate files (#561)
  remotes/origin/nx-infer/5-webpack                                  f395d079 Use inferred tasks for Webpack
  remotes/origin/production                                          cc2389b9 Merge branch 'main' into production
  remotes/origin/renovate/non-major-dependencies                     ac497e6e chore(deps): update non-major dependencies
  remotes/origin/testing                                             a546945d Cherry-pick bug fixes from main (#204)
  remotes/origin/update-now-playing-when-channel-or-playlist-changes 67ffaaf7 wip: Trying to figure this out.
  remotes/origin/user-avatar                                         68b0e61c start upload

(verbose) git rev-parse --verify --abbrev-ref @{-1}
main

(verbose) git remote get-url origin
https://github.com/REDACTED/REDACTED.git

(verbose) git log main..mui/base-ui/9 --format=%s --reverse
mui/material-ui#41031

(verbose) git log main..router --format=%s --reverse
Fix VS Code showing error "Invalid URL"
Add SWC to remove Nx warning
Split router and root providers into separate files

(verbose) git log main..production --format=%s --reverse
Change GET /api/kiosk/devices to POST (#249)
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
deploy
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Change creationDate to be current date in Apple TV JSON
Merge branch 'update-creation-date-to-now' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production
Merge branch 'main' into production

[router] git checkout main
Switched to branch 'main'
Your branch is behind 'origin/main' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)

[main] git rebase origin/main --no-update-refs
Successfully rebased and updated refs/heads/main.

(verbose) git rev-list --left-right main...origin/main

[main] git checkout production
Switched to branch 'production'
Your branch is up to date with 'origin/production'.

[production] git rebase origin/production --no-update-refs
Current branch production is up to date.

(verbose) git rev-list --left-right production...origin/production

(verbose) git show-ref --verify --quiet refs/heads/router

[production] git checkout router
Switched to branch 'router'
Your branch is based on 'origin/router', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)

(verbose) git show-ref --verify --quiet refs/heads/main

(verbose) git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

(verbose) git checkout router
Switched to branch 'router'
Your branch is based on 'origin/router', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)

(verbose) git branch -vva --sort=refname
  main                                                               e0383e6f [origin/main] Split router and root providers into separate files (#561)
  mui/base-ui/9                                                      b3c47700 mui/material-ui#41031
  production                                                         cc2389b9 [origin/production] Merge branch 'main' into production
* router                                                             908abe83 [origin/router: gone] Split router and root providers into separate files
  remotes/origin/HEAD                                                -> origin/main
  remotes/origin/client-frontend-temp                                ac6ee219 temp
  remotes/origin/main                                                e0383e6f Split router and root providers into separate files (#561)
  remotes/origin/nx-infer/5-webpack                                  f395d079 Use inferred tasks for Webpack
  remotes/origin/production                                          cc2389b9 Merge branch 'main' into production
  remotes/origin/renovate/non-major-dependencies                     ac497e6e chore(deps): update non-major dependencies
  remotes/origin/testing                                             a546945d Cherry-pick bug fixes from main (#204)
  remotes/origin/update-now-playing-when-channel-or-playlist-changes 67ffaaf7 wip: Trying to figure this out.
  remotes/origin/user-avatar                                         68b0e61c start upload

(verbose) git config -lz --includes --global
user.name
Stephen Wadeuser.email
REDACTEDpull.rebase
truepull.ff
onlypush.autosetupremote
truediff.colormoved
zebracore.pager
less -x2core.fsmonitor
trueinit.defaultbranch
mainfetch.prune
truerebase.updaterefs
truegit-town.github-token
REDACTEDalias.append
town appendalias.compress
town compressalias.contribute
town contributealias.diff-parent
town diff-parentalias.hack
town hackalias.observe
town observealias.park
town parkalias.prepend
town prependalias.propose
town proposealias.repo
town repoalias.set-parent
town set-parentalias.ship
town shipalias.sync
town syncalias.ls
town branchalias.co
town switchalias.delete
town deletealias.rename
town renamemaintenance.repo
/Users/stephen.wade/git/REDACTED
(verbose) git config -lz --includes --local
core.repositoryformatversion
0core.filemode
truecore.bare
falsecore.logallrefupdates
truecore.ignorecase
truecore.precomposeunicode
truecore.hookspath
.husky/_remote.origin.url
https://github.com/REDACTED/REDACTED.gitremote.origin.fetch
+refs/heads/*:refs/remotes/origin/*branch.main.remote
originbranch.main.merge
refs/heads/mainbranch.main.vscode-merge-base
origin/maingit-town.github-token
REDACTEDgit-town.observed-branches
mui/base-ui/9 type-check-allgit-town.prototype-branches
test iovalkey new-iconifygit-town-branch.mui/base-ui/9.parent
mainlog.excludedecoration
refs/gitbutlerbranch.production.remote
originbranch.production.merge
refs/heads/productionbranch.production.vscode-merge-base
origin/mainmaintenance.auto
falsemaintenance.strategy
incrementalbranch.mui/base-ui/9.vscode-merge-base
origin/maingit-town-branch.router.parent
mainbranch.router.vscode-merge-base
origin/mainbranch.router.remote
originbranch.router.merge
refs/heads/router
(verbose) git stash list
stash@{0}: WIP on (no branch): 54289567 Add test for alwaysMask

Ran 30 shell commands.
```

**Comment by stephenwade** [comment]
The test passes fine, but my own test case fails, and this is the same all the way back in Git Town v10. It doesn't seem to be a regression in Git Town within the last year.

To rule out any possible config issues, I created a brand-new Debian VM with UTM and reproduced my issue in there.

<details>
<summary>Command log</summary>

```
❯ ssh stephen@192.168.4.99
The authenticity of host '192.168.4.99 (192.168.4.99)' can't be established.
ED25519 key fingerprint is SHA256:py0pwWReNdueJOSDf0wTi9/8n5qEglCrMOLE3B77UW4.
This key is not known by any other names.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '192.168.4.99' (ED25519) to the list of known hosts.
stephen@192.168.4.99's password: 
Linux debian-utm 6.1.0-27-arm64 #1 SMP Debian 6.1.115-1 (2024-11-01) aarch64

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Last login: Tue Nov 19 02:21:24 2024
stephen@debian-utm:~$ sudo apt install git -y
[sudo] password for stephen: 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  git-man liberror-perl patch
Suggested packages:
  git-daemon-run | git-daemon-sysvinit git-doc git-email git-gui gitk gitweb
  git-cvs git-mediawiki git-svn ed diffutils-doc
The following NEW packages will be installed:
  git git-man liberror-perl patch
0 upgraded, 4 newly installed, 0 to remove and 0 not upgraded.
Need to get 9,340 kB of archives.
After this operation, 48.8 MB of additional disk space will be used.
Get:1 http://deb.debian.org/debian bookworm/main arm64 liberror-perl all 0.17029-2 [29.0 kB]
Get:2 http://deb.debian.org/debian bookworm/main arm64 git-man all 1:2.39.5-0+deb12u1 [2,054 kB]
Get:3 http://deb.debian.org/debian bookworm/main arm64 git arm64 1:2.39.5-0+deb12u1 [7,137 kB]
Get:4 http://deb.debian.org/debian bookworm/main arm64 patch arm64 2.7.6-7 [121 kB]
Fetched 9,340 kB in 0s (19.1 MB/s)
Selecting previously unselected package liberror-perl.
(Reading database ... 28571 files and directories currently installed.)
Preparing to unpack .../liberror-perl_0.17029-2_all.deb ...
Unpacking liberror-perl (0.17029-2) ...
Selecting previously unselected package git-man.
Preparing to unpack .../git-man_1%3a2.39.5-0+deb12u1_all.deb ...
Unpacking git-man (1:2.39.5-0+deb12u1) ...
Selecting previously unselected package git.
Preparing to unpack .../git_1%3a2.39.5-0+deb12u1_arm64.deb ...
Unpacking git (1:2.39.5-0+deb12u1) ...
Selecting previously unselected package patch.
Preparing to unpack .../patch_2.7.6-7_arm64.deb ...
Unpacking patch (2.7.6-7) ...
Setting up liberror-perl (0.17029-2) ...
Setting up patch (2.7.6-7) ...
Setting up git-man (1:2.39.5-0+deb12u1) ...
Setting up git (1:2.39.5-0+deb12u1) ...
Processing triggers for man-db (2.11.2-2) ...
stephen@debian-utm:~$ curl https://www.git-town.com/install.sh | sh
-bash: curl: command not found
stephen@debian-utm:~$ sudo apt install curl -y
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libcurl4
The following NEW packages will be installed:
  curl libcurl4
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 674 kB of archives.
After this operation, 1,428 kB of additional disk space will be used.
Get:1 http://deb.debian.org/debian bookworm/main arm64 libcurl4 arm64 7.88.1-10+deb12u8 [365 kB]
Get:2 http://deb.debian.org/debian bookworm/main arm64 curl arm64 7.88.1-10+deb12u8 [309 kB]
Fetched 674 kB in 0s (3,001 kB/s)
Selecting previously unselected package libcurl4:arm64.
(Reading database ... 29690 files and directories currently installed.)
Preparing to unpack .../libcurl4_7.88.1-10+deb12u8_arm64.deb ...
Unpacking libcurl4:arm64 (7.88.1-10+deb12u8) ...
Selecting previously unselected package curl.
Preparing to unpack .../curl_7.88.1-10+deb12u8_arm64.deb ...
Unpacking curl (7.88.1-10+deb12u8) ...
Setting up libcurl4:arm64 (7.88.1-10+deb12u8) ...
Setting up curl (7.88.1-10+deb12u8) ...
Processing triggers for man-db (2.11.2-2) ...
Processing triggers for libc-bin (2.36-9+deb12u9) ...
stephen@debian-utm:~$ curl https://www.git-town.com/install.sh | sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4230  100  4230    0     0  10340      0 --:--:-- --:--:-- --:--:-- 10367
GIT TOWN INSTALLER SCRIPT

This installer is under development. Please report bugs at
https://github.com/git-town/git-town/issues/new.

No other Git Town installation found, proceeding to install Git Town 16.6.0 for linux_arm_64.

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 5086k  100 5086k    0     0  5926k      0 --:--:-- --:--:-- --:--:-- 5926k

I have successfully:
- downloaded https://github.com/git-town/git-town/releases/download/v16.6.0/git-town_linux_arm_64.tar.gz
- extracted the Git Town 16.6.0 binary to /home/stephen/.local/bin/git-town

Please perform these steps to finalize the installation:
- add /home/stephen/.local/bin to the PATH
stephen@debian-utm:~$ chmod +x .local/bin/git-town
stephen@debian-utm:~$ source .profile
stephen@debian-utm:~$ cd $(mktemp -d)
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git init
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint: 
hint: 	git config --global init.defaultBranch <name>
hint: 
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint: 
hint: 	git branch -m <name>
Initialized empty Git repository in /tmp/tmp.iv2u3UkV4x/.git/
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git commit --allow-empty -m 'Initial commit'
Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'stephen@debian-utm.(none)')
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git config --global user.email "stephen@stephenwade.me"
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git config --global user.name "Stephen Wade"
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git commit --allow-empty -m 'Initial commit'
[master (root-commit) 5346ce6] Initial commit
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git town config setup
Aliased commands: (none)
Main branch: master
Perennial regex: 
Default branch type: feature
Feature regex: 
Code hosting: auto-detect
Origin hostname: (not provided)
Sync feature branches: merge
Sync perennial branches: rebase
Sync with upstream: yes
Sync tags: yes
Push new branches: no
Push hook: enabled
Create prototype branches: disabled
Ship deletes tracking branches: api
Ship deletes tracking branches: yes
Config storage: file

stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git town hack feature

[master] git add -A

[master] git stash
Saved working directory and index state WIP on master: 5346ce6 Initial commit

[master] git checkout -b feature
Switched to a new branch 'feature'

[feature] git stash pop
On branch feature
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   .git-branches.toml

Dropped refs/stash@{0} (5775ee9d8fff5d092a4672e75b7436daa311b2fc)


branch "feature" is now a child of "master"
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ touch feature.txt
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git add .
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git commit -m 'Add feature'
[feature c3d4c2d] Add feature
 2 files changed, 112 insertions(+)
 create mode 100644 .git-branches.toml
 create mode 100644 feature.txt
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git switch -
Switched to branch 'master'
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git merge feature --squash
Updating 5346ce6..c3d4c2d
Fast-forward
Squash commit -- not updating HEAD
 .git-branches.toml | 112 +++++++++++++++++++++++++++++++++++++++++++++++++++++
 feature.txt        |   0
 2 files changed, 112 insertions(+)
 create mode 100644 .git-branches.toml
 create mode 100644 feature.txt
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git commit -m 'Merge feature'
[master d982cb7] Merge feature
 2 files changed, 112 insertions(+)
 create mode 100644 .git-branches.toml
 create mode 100644 feature.txt
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git town sync --all --verbose
Git Town 16.6.0

(verbose) git version
git version 2.39.5

(verbose) git rev-parse --show-toplevel
/tmp/tmp.iv2u3UkV4x

(verbose) git config -lz --includes --global
user.email
stephen@stephenwade.meuser.name
Stephen Wade
(verbose) git config -lz --includes --local
core.repositoryformatversion
0core.filemode
truecore.bare
falsecore.logallrefupdates
truegit-town-branch.feature.parent
master
(verbose) git branch -vva --sort=refname
  feature c3d4c2d Add feature
* master  d982cb7 Merge feature

(verbose) git status --long --ignore-submodules
On branch master
nothing to commit, working tree clean

(verbose) git remote

(verbose) git stash list

(verbose) git branch -vva --sort=refname
  feature c3d4c2d Add feature
* master  d982cb7 Merge feature

(verbose) git rev-parse --verify --abbrev-ref @{-1}
feature

(verbose) git remote get-url origin
error: No such remote 'origin'

(verbose) git log master..feature --format=%s --reverse
Add feature

[master] git checkout feature
Switched to branch 'feature'

[feature] git merge --no-edit --ff master
Merge made by the 'ort' strategy.

(verbose) git show-ref --verify --quiet refs/heads/master

[feature] git checkout master
Switched to branch 'master'

(verbose) git show-ref --verify --quiet refs/heads/feature

(verbose) git branch -vva --sort=refname
  feature 8c8776d Merge branch 'master' into feature
* master  d982cb7 Merge feature

(verbose) git config -lz --includes --global
user.email
stephen@stephenwade.meuser.name
Stephen Wade
(verbose) git config -lz --includes --local
core.repositoryformatversion
0core.filemode
truecore.bare
falsecore.logallrefupdates
truegit-town-branch.feature.parent
master
(verbose) git stash list

Ran 21 shell commands.stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ git branch
  feature
* master
stephen@debian-utm:/tmp/tmp.iv2u3UkV4x$ 
```

</details>

@kevgo Any idea why my test case would be failing?

Here's a cleaned up list of commands that produce the same result:

```
sudo apt install curl git -y
curl https://www.git-town.com/install.sh | sh
chmod +x .local/bin/git-town
source .profile
cd $(mktemp -d)
git init
printf '[branches]\nmain = "master"\n' > .git-branches.toml
git add .
git config --global user.email "stephen@stephenwade.me"
git config --global user.name "Stephen Wade"
git commit -m 'Initial commit'
git town hack feature
touch feature.txt
git add .
git commit -m 'Add feature'
git switch -
git merge feature --squash
git commit -m 'Merge feature'
git town sync --all --verbose
git branch
```

**Comment by stephenwade** [comment]
Yes, it does.

```
2 scenarios (2 passed)
34 steps (34 passed)
255.075134ms
ok      github.com/git-town/git-town/v16        0.317s
```

Can you reproduce the issue with my commands?

**Comment by kevgo** [comment]
I noticed in your repro at https://github.com/git-town/git-town/issues/4222#issuecomment-2484913053 that you create a local branch `feature`, squash-merge its commits into the main branch, and then run `git sync --all` on the main branch. In this scenario, `git sync` should not delete the feature branch. Local branches are only deleted if their tracking branch has been removed from the remote, which isn't happening here. Could you clarify why you expect a different behavior in this case?

The log in https://github.com/git-town/git-town/issues/4222#issuecomment-2484767973 shows a situation where the tracking branch of branch `router` is indeed deleted, and `git town sync` doesn't delete it. I'll keep thinking about this one.

**Comment by stephenwade** [comment]
> Local branches are only deleted if their tracking branch has been removed from the remote, which isn't happening here. Could you clarify why you expect a different behavior in this case?

Thank you for explaining the behavior! I didn't realize Git Town was looking at the remote. I thought it was checking to see if the contents of the feature branch was included in the main branch.

When it happens to me in the real world, the branch is indeed gone from the remote. I'll make my own test case to reproduce that and try again to bisect.

**Comment by kevgo** [comment]
The ability to automatically delete "empty" branches is definitely useful from time to time. There is a [plan](https://github.com/git-town/git-town/issues/2557) for it, but it's not implemented yet.

**Comment by stephenwade** [comment]
I was able to reproduce the problem in my Debian VM. I used a second local repository and added it as a remote: https://stackoverflow.com/questions/10603671/how-to-add-a-local-repo-and-treat-it-as-a-remote-repo.

I also found that with this script, it behaved correctly 6 months ago.

I bisected using the script below and found that the problem was introduced in https://github.com/git-town/git-town/pull/4127.

Here's the script I used (with `git bisect run`).

```sh
make install || make build  # The correct command to use changed during the period I am testing
cd $(mktemp -d)
# Create a new repo
git init
printf '[branches]\nmain = "main"\n' > .git-branches.toml
git add .
git commit -m "Initial commit with $(git-town --version)"
# Add a remote
REMOTE=$(mktemp -d)
git init --bare $REMOTE
git remote add origin $REMOTE
# Add a feature branch
git-town hack feature
touch feature.txt
git add .
git commit -m "Add feature.txt"
git-town sync -a  # Sync the feature branch to the remote
git checkout main
git merge feature --squash
git commit -m "Merge feature branch"
git push origin :feature  # Delete the `feature` branch on the remote
git-town sync -a
! git rev-parse --verify --quiet feature > /dev/null
exit $?  # Exit with an error code if the `feature` branch still exists
```

**Comment by kevgo** [comment]
Excellent work, and thanks so much for digging into this! 🙏

I was able to reproduce the failure using your test script. It passes at commit `972e66e` but starts failing at commit `1a2fe49`, just as you pointed out.

What’s odd is that your script seems to follow roughly the same steps as my end-to-end test, yet my test doesn’t trigger the bug. I'll look into that as well.

**Comment by kevgo** [comment]
I have shipped a bug fix that makes your test script pass. It's not released yet, but you can try it out today by running:

```
git checkout main
git pull
make install
```

Then use the compiled Git Town executable normally.

If the issue persists, please reopen this ticket. Thanks!

**Comment by kevgo** [comment]
This is released in [v16.7.0](https://github.com/git-town/git-town/releases/tag/v16.7.0).
